### ISSUE TITLE
Fixtures without name matching magic

### ISSUE THREAD

**Issue by shoyer** [pytest-dev/pytest#3834]
I've been happily using pytest for several projects for the past few years.

There's one part about pytest that I still struggle to get behind: The way that fixtures magically match argument names to fixtures -- and apparently [I'm not alone in this feeling](https://mail.python.org/pipermail/numpy-discussion/2018-August/078560.html). I would much rather declare dependencies explicitly in some way using code. I know this would be more verbose, but that's a tradeoff I'm happy to make.

Is it possible to do this in some way with pytest today? If not, would you be open to adding an  *optional* feature for this?

I was thinking perhaps something like the following, using an example adapted from [the docs](https://docs.pytest.org/en/latest/fixture.html):
```python
import pytest

@pytest.fixture
def smtp_connection():
    import smtplib
    return smtplib.SMTP("smtp.gmail.com", 587, timeout=5)

@pytest.use_fixture(stmp_connection)
def test_ehlo(connection):
    response, msg = connection.ehlo()
    assert response == 250
    assert 0 # for demo purposes
```

**Comment by nicoddemus** [comment]
Hi @shoyer,

If you don't like the way fixtures are automatically injected, you can be more explicit in a way very similar to the example you provided:

```python
import pytest

@pytest.fixture
def smtp_connection():
    import smtplib
    return smtplib.SMTP("smtp.gmail.com", 587, timeout=5)

@pytest.mark.usefixtures("stmp_connection")
def test_ehlo(smtp_connection):
    response, msg = connection.ehlo()
    assert response == 250
    assert 0 # for demo purposes
```

**Comment by shoyer** [comment]
@nicoddemus that's explicit about using a fixture, but it's still using a string to refer to a name variable. Is it possible to do something like that while referencing `smtp_connection` as a Python object -- preferably with a different, non-conflicting argument name?

**Comment by nicoddemus** [comment]
Probably you can do something like this:

```python
import pytest

def use_fixture(fixfunc):
    def inner(f):
        return pytest.mark.usefixtures(fixfunc.__name__)(f)
    return inner

@pytest.fixture
def fix():
    return 1

@use_fixture(fix)
def test(fix):
    assert fix == 1
```

But you still need to use the same name in the argument as the fixture, plus you will need to explicitly import the fixture function into the test module in case it is not defined there, and that is not really recommended because it messes with the internal caching of the fixtures.

> that's explicit about using a fixture, but it's still using a string to refer to a name variable.

What is the problem with that in your opinion? That's very explicit IMHO: `@pytest.mark.usefixtures("fix")` explicitly declares that the test function uses a fixture `"fix"`, and that is passed as argument name.

> preferably with a different, non-conflicting argument name?

Not sure what is the reason for that, can you explain?

**Comment by shoyer** [comment]
> But you still need to use the same name in the argument as the fixture, plus you will need to explicitly import the fixture function into the test module in case it is not defined there, and that is not really recommended because it messes with the internal caching of the fixtures.

Indeed, but also using `fixfunc.__name__` for lookup is a good example of exactly what I was trying to avoid here.

> What is the problem with that in your opinion? That's very explicit IMHO: `@pytest.mark.usefixtures("fix")` explicitly declares that the test function uses a fixture "fix", and that is passed as argument name.

I want to use Python's normal rules for resolving symbols (i.e., standard variable lookup), and I want to keep introspection to the bare minimum needed to run tests. I know pytest uses magic to rewrite `assert` and decide which tests to run, but in general I like my Python code (even tests) to follow the normal rules of Python code.

Referencing functions by string name instead of object breaks normal code introspection tools (e.g., lint), which can't recognize misspelled names if there's no variable reference. So if I write `stmp_connection` instead of `smtp_connection`, nothing catches that mistake until I actually run my tests.

(In my opinion, one of the best things about Python is that dependencies are explicit, unless you use the discouraged `import *`.)

>> preferably with a different, non-conflicting argument name?
> Not sure what is the reason for that, can you explain?

As one simple example, if you run this example through `pylint`, you get the following warning:
```
W0621: Redefining name 'smtp_connection' from outer scope (line 7) (redefined-outer-name)
```

At Google, we actually [strictly enforce running pylint](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#21-lint) on all Python code. If I wanted to disable this message, I would need to write a comment `# pylint: disable=redefined-outer-name`.

Now, pylint is known for being overly-aggressive in some cases, but I do think that in general it's best not mask variable names from outer scopes. There's basically no reason why this is ever necessary if not using pytest, and if done unintentionally, this can easily lead to bugs.

-------------------

Another example of what doing this explicitly could look like, with keyword arguments:
```python
@pytest.mark.usefixtures(conn=stmp_connection)
def test_ehlo(conn):
    ...
```

Using a fixture imported from another module also becomes very straightforward, e.g.,
```python
@pytest.mark.usefixtures(conn=other_module.stmp_connection)
def test_ehlo(conn):
    ...
```

**Comment by nicoddemus** [comment]
Hi @shoyer sorry for taking so long to get back at this.

I'm not sure it has a place in the core though, as it would lead to a different way of doing things, and is solving not really an existing problem but a personal preference which is shared by some people but not by the majority of pytest users (at least that's what I think, no hard data here).

But all you are proposing is probably possible to write as a plugin. I suggest you give this a try (be sure to post here if you need help) and see how that feels in real code.

What do you think?

**Comment by burnpanck** [comment]
Let me add to this that the name-matching *does* lead to real-world problems: I just spent an hour trying to get my pytest-sanic tests to working, not realising that it's `test_client` fixture clashes with the `text_client` fixture of pytest-aiohttp.

**Comment by tuukkamustonen** [comment]
This would be really nice. I hate dropping `# pylint: disable=redefined-outer-name` lines to my test files because `pylint` will nag about every test with a fixture otherwise.

The original suggestion by @shoyer looks good enough, imo. Anyway, here are some extra thoughts:

```python
# not supported by python(?):

def test_ehlo(@pytest.inject(smtp_connection) connection):
    ...

# dunno if this would be doable, either:

@pytest.fixture
class SmtpConnection:
    def __call__(self):
        import smtplib
        return smtplib.SMTP("smtp.gmail.com", 587, timeout=5)

@pytest.inject
def test_ehlo(conn: SmtpConnection):
    ...
```

**Comment by nicoddemus** [comment]
> This would be really nice. I hate dropping # pylint: disable=redefined-outer-name lines to my test files because pylint will nag about every test with a fixture otherwise.

Just to mention that fixtures have a `name` parameter to avoid this problem:

```python
@pytest.fixture(name='smtp_connection')
def smtp_connection_():
    ...

def test_ehlo(smtp_connection):
    ...
```

**Comment by tuukkamustonen** [comment]
Oh, that's a workaround I could use. Good to know.

Though the benefit in being able to do something like what I suggested above would be that when you (have to) import fixtures, they are then actually referenced in the tests file, and your import will not get marked as unused (and removed if you auto-organize imports). Sure I could:

```
from foo import my_fixture

PYTEST_IMPORTS = my_fixture
```

But then `PYTEST_IMPORTS` will be marked as unused :P

**Comment by countvajhula** [comment]
I'll register my support for this issue as well. When users first see tests written in pytest, it is mystifying how the arguments to these functions get populated. At least having the option to be explicit about it in a robust way would be nice.

Another possible benefit is the ability to give a different name to the same fixture in different tests. Or maybe more specifically, the ability to give the same name to different fixtures in different tests. Not sure if this is already possible, but with an explicit mechanism like the one @shoyer is talking about, it could look something like this:

Let's say we have these fixtures for an application that processes log files (this is a barebones example just to illustrate the point):
```
@pytest.fixture
def simple_logfile():
    return ("2019-08-16 10:35:05 connection established\n"
             "2019-08-16 11:07:16 connection dropped\n")

@pytest.fixture
def empty_logfile():
    return ""
```

With the current behavior, we could write tests like this:
```
def test_read_logfile(simple_logfile):
    result = read_logfile(simple_logfile)
    assert result == True

def test_read_empty_logfile(empty_logfile):
    result = read_logfile(empty_logfile)
    assert result == True
```

In the proposed behavior, it could look like this:
```
@pytest.mark.usefixtures(logfile=simple_logfile)
def test_read_logfile(logfile):
    result = read_logfile(logfile)
    assert result == True

@pytest.mark.usefixtures(logfile=empty_logfile)
def test_read_empty_logfile(logfile):
    result = read_logfile(logfile)
    assert result == True
```

The explicit approach allows the generic aspects of the test code to remain generic -- in this example, the test only cares about reading a logfile and asserting that it works. The fact that we have selected this file to be an empty one to test an edge case is already captured in the name of the function, and avoiding the need for different variable names seems desirable since the code is the same.

In fact, doing it this way gives us yet another advantage -- the possibility of parametrizing over fixtures using `pytest.mark.parametrize`. For instance, the proposed behavior could now look like this:

```
@pytest.mark.parametrize(
    "logfile", (simple_logfile, empty_logfile),
)
def test_read_logfile(logfile):
    result = read_logfile(logfile)
    assert result == True
```
... which covers both tests in a single parameterized test.

It looks like there is currently already a way to [parametrize fixtures themselves](https://docs.pytest.org/en/latest/fixture.html#parametrizing-fixtures). That's a great feature as well, but I think the ability to parameterize the _test_ rather than the _fixture_ is a different thing which could complement the current functionality and lend additional flexibility.

**Comment by jadkik** [comment]
> I think the ability to parameterize the test rather than the fixture is a different thing which could complement the current functionality and lend additional flexibility.

FYI there is already to do this in a nice-ish way:

```python
import pytest


@pytest.fixture
def simple_logfile():
    return 'simple_logfile contents'


@pytest.fixture
def empty_logfile():
    return 'empty_logfile contents'


@pytest.fixture(name='logfile')
def dynamic_fixture(request):
    return request.getfixturevalue(request.param)


@pytest.mark.parametrize('logfile', ['simple_logfile', 'empty_logfile'], indirect=True)
def test_something(logfile):
    print()
    print('test_something', repr(logfile))
```

Downside being that you can't do indirect parameterization of `simple_logfile` from `test_something`, and it doesn't show up when you run pytest with `--setup-plan` (it does when you do `--setup-only` though.

I can still see the benefit in this proposal, though: if `usefixtures` accepts keyword arguments with names or actual fixture functions, it could be a change that doesn't break the current behaviour.

**Comment by ms-lolo** [comment]
Just wanted to drop my support for this request as well. The current fixtures functionality acts as a global registry of magical variable names and globals are confusing and error prone for the reasons already outlined a few times above. When I'm creating new fixtures I should not need to try and think of a globally unique name for it as long as I've implemented it in my globally unique package name.

It's a little disappointing to try and convince folks on the team to write maintainable/testable code and then run into the testing framework itself not following software engineering best practices. We only need to write this code once but we have to read it, understand it, and maintain it for years. We can afford to type a few extra lines of code to make sure it works and is immediately clear to anyone that might run into it in the future.

**Comment by nicoddemus** [comment]
Hi @ms-lolo, 

> It's a little disappointing to try and convince folks on the team to write maintainable/testable code and then run into the testing framework itself not following software engineering best practices. We only need to write this code once but we have to read it, understand it, and maintain it for years. We can afford to type a few extra lines of code to make sure it works and is immediately clear to anyone that might run into it in the future.

I sympathize with that sentiment, but personally I think the issue boils down more to introducing another way to use fixtures, which might lead to even more confusion when the classic way and this new way proposed here are mixed. 

As was stated before, I believe it is possible to implement a different mechanism as a plugin, but adding this directly to the core without first seeing it being used in the wild might not be the best approach.

**Comment by shoyer** [comment]
> I sympathize with that sentiment, but personally I think the issue boils down more to introducing another way to use fixtures, which might lead to even more confusion when the classic way and this new way proposed here are mixed.

If we agree that an explicit API is better, we might consider deprecating the current global name matching entirely, and eventually only supporting the new way (or at least suggesting it as the preferred option for new code).

**Comment by ms-lolo** [comment]
Totally understand wanting to avoid having multiple ways of doing this. And using a plugin to demonstrate a cleaner approach sounds like a great idea that I'm going to try and find some time to investigate, although I have no experience with the pytest internals, so any guidance would be appreciated!

> I'm not sure it has a place in the core though, as it would lead to a different way of doing things, and is solving not really an existing problem but a personal preference which is shared by some people but not by the majority of pytest users (at least that's what I think, no hard data here).

I was mostly leaving a comment to add a data point about this comment :) I disagree with the statement that this is a personal preference but we don't need to go down that rabbit hole without even a proof of concept alternative to compare to.

**Comment by countvajhula** [comment]
@jadkik Nice, though another downside is that, judging by your example, we'd need to write a wrapping dynamic fixture for every set of fixtures we might want to parametrize over. And since we would be creating this abstraction behind yet another global name, this does not escape the namespace issues that @ms-lolo brings up. By separating the fixture definition from its name, we gain the ability to abstract over any set of fixtures behind a lexically-scoped name in each test, without having to pre-designate (via a wrapper function like `dynamic_fixture`) which ones those are.

**Comment by The-Compiler** [comment]
FWIW maybe we could turn to [ward](https://wardpy.com/) for inspiration here. It [uses default argument values](https://wardpy.com/guide/fixtures/) to make fixtures more explicit (and namespaced) without having much more boilerplate:

```python
from ward import fixture, test
@fixture
def user():
    return User(id=1, name="sam")

@test("fetch_user_by_id should return the expected User object")
def _(expected_user=user):
    fetched_user = fetch_user_by_id(id=expected_user.id)
    assert fetched_user == expected_user
```

I don't particularly like the `@test` decorator/naming thingy myself, but having namespaced fixtures which can be imported and used by specifying them as argument value seems pretty nice to me.

Of course the million dollar question is how to implement this with the already quite complex fixture code, especially in a way that's backwards compatible. So far I haven't had the time (or mind-space) to see if it'd even be realistic at all.

**Comment by RonnyPfannschmidt** [comment]
anything in the domain of adoption/modifying the mechanisms in which fixture lookup works is something for after we de-cruft and de-tangle it

from past experience its pretty much doomed before we simplify/expliticify it

my take is that we can't hope to do better until we make it reasonably sane to choose the dependency injection system and to integrate a external one with pytest fixtures as a base

personally i really want to enable better lookup, but that comes after de-tangleing what grew due to the history that started with the `pytest_funcarg__somename` hooks that wouldn't cache

**Comment by graingert** [comment]
now that pytest is Python3.5+ only, using a Annotations based DI is possible: https://fastapi.tiangolo.com/tutorial/dependencies/

**Comment by nicoddemus** [comment]
@shoyer

> If we agree that an explicit API is better, we might consider deprecating the current global name matching entirely

That's an idea, and probably the way to go for certain features, but not for fixture IMO: it would break every pytest suite out there which uses fixtures. This kind of breakage would kill the project IMO.


@ms-lolo

> Totally understand wanting to avoid having multiple ways of doing this. And using a plugin to demonstrate a cleaner approach sounds like a great idea that I'm going to try and find some time to investigate, although I have no experience with the pytest internals, so any guidance would be appreciated!

It depends on the mechanism actually... there are some proposals in this thread, but importing and using the actual names in the test function definition, bypassing the current global lookup, might not be possible through a simple plugin, I'm afraid.

@RonnyPfannschmidt

> my take is that we can't hope to do better until we make it reasonably sane to choose the dependency injection system and to integrate a external one with pytest fixtures as a base

Indeed, but it seems this would take many hours of someone dedicated to this task to tackle. 

> now that pytest is Python3.5+ only, using an Annotations based DI is possible: fastapi.tiangolo.com/tutorial/dependencies

Indeed, and that's similar to what `ward` does, as @The-Compiler commented.

---

My take overall:

* Enabling more explicit fixture lookup would be great
* Currently the pytest interanals don't allow for that
* Backward compatibility is a hard requirement

So this is a very complicated issue.

**Comment by graingert** [comment]
> Backward compatibility is a hard requirement

is backwards and forwards compatibility a hard requirement? Can tests using sans-name-magic based fixtures be wholly incompatible with name-magic based fixtures, specifically a test requiring Annotation based fixtures would not be able to consume name-based fixtures (except autouse?)?

**Comment by nicoddemus** [comment]
> Can tests using sans-name-magic based fixtures be wholly incompatible with name-magic based fixtures, specifically a test requiring Annotation based fixtures would not be able to consume name-based fixtures (except autouse?)?

That's a good question. I suppose both systems need to co-exist and be interchangeable, so a test/fixture can request fixtures using name matching and/or explicit annotations, otherwise it wouldn't be very useful.

**Comment by techdragon** [comment]
@nicoddemus "co-existence" in that they are both functional options you can use as a developer, in the same version of pytest, sounds like a good thing to have for compatibility reasons... 

But "interchangeable" actually sounds like a bad thing to me... it sounds like it would make building this more complicated than I feel it needs to be. I would be happy switching over to explicit fixtures if I had to use some special compatibility shim in my own codebase as a wrapper around magic fixtures. I kind of already have to, using things similar to the examples by yourself https://github.com/pytest-dev/pytest/issues/3834#issuecomment-428923623 and @jadkik https://github.com/pytest-dev/pytest/issues/3834#issuecomment-581358040 if I want to keep the level of magic in my tests to a minimum.

**Comment by RonnyPfannschmidt** [comment]
i believe we could have

```
@fixture(name=None, typed=True)
def something(request) -> MyClass
   return MyClass(name=request.fixturename)


def test_has_some(alice: MyClass, bob: MyClass):
   assert alice.name == "alice"
  assert bob.name == "bob"
```
in future, but the timeline is absolutely unclear

**Comment by ms-lolo** [comment]
> i believe we could have
> 
> ```python
> @fixture(name=None, typed=True)
> def something(request) -> MyClass
>    return MyClass(name=request.fixturename)
> 
> 
> def test_has_some(alice: MyClass, bob: MyClass):
>    assert alice.name == "alice"
>   assert bob.name == "bob"
> ```
> 
> in future, but the timeline is absolutely unclear

I'm not quite following the example. How id `test_has_some()` defining where the two parameters come from?

**Comment by graingert** [comment]
@RonnyPfannschmidt I'm against looking at the args/kwarg names completely, annotated fixtures should use [typing.Annotated pep 593](https://www.python.org/dev/peps/pep-0593/) which have been reserved for runtime introspection:

```
class MyClass(pytest.Fixture):
    def __init__(self, name: str):
        self.name = name

def test_has_some(a: pytest.Param[MyClass, "alice], b: pytest.Param(MyClass, "bob")):
    assert a.name == "alice"
    assert b.name == "bob"
```

**Comment by RonnyPfannschmidt** [comment]
@graingert i can understand the sentiment, but the ux looks painful at first glance

**Comment by max-sixty** [comment]
Could I make a suggestion that's very close to the original one by @shoyer ?

Two modes — explicit & implicit — default is for both to work. A pytest config is available to enforce explicit only (like how `--strict-markers` works now).
- Implicit — the existing design
- Explicit — something like:

```python
import pytest

@pytest.fixture
def connection():
    import smtplib
    return smtplib.SMTP("smtp.gmail.com", 587, timeout=5)

@pytest.use_fixture(connection=connection)  # <- not required in implicit mode, would achieve same result without it
def test_ehlo(connection):
    response, msg = connection.ehlo()
    assert response == 250
    assert 0 # for demo purposes
```
Note the annotated line can be elided in implicit mode because the name `connection` is used as both the fixture name and the parameter name.

In explicit mode, if we want to reduce the number of times `connection` is overloaded, we can use a different name for the function name and parameter name.
```python
import pytest

@pytest.fixture
def smtp_connection():
    import smtplib
    return smtplib.SMTP("smtp.gmail.com", 587, timeout=5)

@pytest.use_fixture(connection=smtp_connection)  # <- required, would not work in implicit mode given the name difference
def test_ehlo(connection):
    response, msg = connection.ehlo()
    assert response == 250
    assert 0 # for demo purposes
```

**Comment by ms-lolo** [comment]
@max-sixty that's pretty much what I was hoping for. It's extremely clear and obvious and the function itself is not coupled to the implementation details of the testing framework. In theory, `test_ehlo` can be defined without the decorator and it would be a completely normal testing function that can be wrapped in `pytest.use_fixture()` later.

**Comment by nicoddemus** [comment]
@max-sixty

> Implicit — the existing design
> Explicit — something like:

Not sure a config file which enables one behavior but disables the other is feasible. For example, some internal pytest features and many plugins depend on the implicit behavior. Aside from the internals (which could be work around), enabling "explicit" means that you can't use any of the existing plugins?

@graingert 
> @RonnyPfannschmidt I'm against looking at the args/kwarg names completely, annotated fixtures should use typing.Annotated pep 593 which have been reserved for runtime introspection

Not sure I follow what you mean here, but I read @RonnyPfannschmidt's example:

```python
@fixture(name=None, typed=True)
def something(request) -> MyClass
    return MyClass(name=request.fixturename)


def test_has_some(alice: MyClass, bob: MyClass):
    assert alice.name == "alice"
    assert bob.name == "bob"
```

As meaning that for each type-annotated parameter, pytest should look at the *type*, and see in the current module namespace if any `@pytest.fixture`-decorated function (either defined locally or imported) has the return value annotated with that type; in that case, call the function and inject the value.

At first glance seems like it could work, but after some thought:

1. This means you can have multiple values for the fixture in the same scope/call. Right now each fixture name has a single value at that scope (for example, you only have a `tmpdir` instance, regardless if the test function and or multiple fixtures request it), but this now is not so clear:

   
   ```python
   @fixture(name=None, typed=True)
   def something(request) -> MyClass
       ...
   
   @fixture(name=None, typed=True)
   def other_fixture(alice: MyClass) -> Foo:
       ...  

   def test_has_some(alice: MyClass, other: Foo):
       ...
   ``` 


   In that case, how many instances of `MyClass` are created? The short answer would be 2 (one for `other_fixture`, another instance to `test_has_some`), but that's different on how normal fixtures work, which is confusing. Making they share the instance is also tricky, how to you define what ties them together? The attribute name? Seems brittle.


2. Right now people are already using this notation for type-annotations (see `pytest-mock` which recently exported `MockerFixture` for this purpose, and also how we plan to proceed with builtin fixtures in pytest). Hijacking it at this point seems problematic to me.

3. Using `typing.Annotated` seems the standard way to accomplish this, but I agree it feels clunky.


---

IMHO a proposal needs to think on the big picture, and how the new way to specify fixtures will interact with the classical way. I think it would be a mistake to make code bases choose one while disabling the other.

**Comment by altendky** [comment]
A type-based lookup doesn't seem great.  We only get one `int` typed fixture?  We have to create not-quite-`int` types to differentiate them while still hinting `int` to mypy and editors?  I thought the solution to the 'magic' was to just use Python to identify the fixture as in `mypackage.somemodule.the_fixture`.

Then we go and add in multiple simultaneous instances of a fixture and...  we're back to some magic for 'which instance do you want?'  :]

**Comment by shoyer** [comment]
> A type-based lookup doesn't seem great. We only get one `int` typed fixture? We have to create not-quite-`int` types to differentiate them while still hinting `int` to mypy and editors? I thought the solution to the 'magic' was to just use Python to identify the fixture as in `mypackage.somemodule.the_fixture`.

Agreed, type-based lookup as described in https://github.com/pytest-dev/pytest/issues/3834#issuecomment-681794705 feels like a solution to a different problem, unless the type annotation includes a Python reference to the fixture function.

**Comment by ms-lolo** [comment]
I agree. Anything trying to look at the test function arguments (names or types) sounds like a step in the wrong direction and just another version of implicit fixtures. It becomes especially confusing when mixed with things like the standard `unittest.mock` libraries which works by decorating functions. What @max-sixty suggested seems the least magical and the ability to enable/disable the two types of fixtures seems unrelated.

**Comment by altendky** [comment]
I like click but I hesitate on the decorator route for new development at this point.  What are the benefits over leveraging annotations?

**Comment by max-sixty** [comment]
> Not sure a config file which enables one behavior but disables the other is feasible. For example, some internal pytest features and many plugins depend on the implicit behavior. Aside from the internals (which could be work around), enabling "explicit" means that you can't use any of the existing plugins?

If we want:
a) Backward compatibility
b) Ability to enforce these explicit / non-magical fixtures in a project
...then I think we need some sort of config

The config wouldn't need to apply to plugins — e.g. maybe it's only on the fixtures in that project. 

`markers --strict` is similar IIUC?

**Comment by nicoddemus** [comment]
> b) Ability to enforce these explicit / non-magical fixtures in a project

Oh I didn't really realized this was a requirement. My understanding was to allow explicit fixture, not enforce them on a given project. Note that we don't need to get stuck on this as a requirement; I think the important thing to focus on is to provide the capability of using explicit fixtures, enforcing them is secondary (it can only be even enforced by a linter for example).

> The config wouldn't need to apply to plugins — e.g. maybe it's only on the fixtures in that project.

Should be possible, but not how things work internally.

> markers --strict is similar IIUC?

Not really, it applies to all marks, so it will complain even by markers used by plugins but are not declared explicitly.

**Comment by graingert** [comment]
> but I agree it feels clunky.

Annotated supports arbitrary nesting and aliasing, users wouldn't actually see `typing.Annotated` in their code

> A type-based lookup doesn't seem great. We only get one int typed fixture? We have to create not-quite-int types to differentiate them while still hinting int to mypy and editors?

Again typing.Annoated would work here to runtime annotate an int factory. I'd probably go for a class though:

```
@pytest.Fixture
@attrs.define
class EphemeralPortReserve:
   v: int
```

```
def test_webserver(p: EphemeralPortReserve):
    django.ruinserver(port=p.v)
```
d

**Comment by techdragon** [comment]
@nicoddemus the ability to enforce them was kind of where I was going with the idea that interoperability was potentially an undesirable feature. Even if it meant plugins didn’t work for a while... As long as there’s a simple enough way to make fixtures work both ways (explicit and old style implicit), there’s a lot of people like myself who will step up and start submitting fixes to the libraries that provide fixtures. 

@graingert i get It, but that took quite a lot longer to parse in my head than the simple example from @max-sixty. It also requires runtime type introspection which is still (at least last time I looked) a bit of a risky area as far as performance goes. As a hypothesis user, who is imagining this inspection overhead multiplied by a hundred or a thousand times for every test I write, I’m very concerned about anything that could make that slower. 

Also @graingert in my head I can’t read they type hint example without simplifying it down to some kind of an `@inject` decorator that accepts `*args` and `**kwargs` of callable objects. The type hinting just doesn’t seem to me, to add enough other than looking slightly more modern since I also have started to associate type hints with more modern looking python code. Do you have larger/non-trivial examples that might be more compelling?

**Comment by RonnyPfannschmidt** [comment]
By now im strongly opposed to use type hints for parameter configuration, it's just terrifying complex to read

I'd rather use a decorator to pass over the di configuration

Until we enable to actually do it, i don't see value in hypothesis on how it will look, as implementation and hitting issues will reshape the system.

**Comment by nicoddemus** [comment]
> @nicoddemus the ability to enforce them was kind of where I was going with the idea that interoperability was potentially an undesirable feature. Even if it meant plugins didn’t work for a while... As long as there’s a simple enough way to make fixtures work both ways (explicit and old style implicit), there’s a lot of people like myself who will step up and start submitting fixes to the libraries that provide fixtures.

I see, but I don't think we need to enforce this, I suppose it is possible to make the architecture flexible enough to allow for the styles to be mixed *or* enforced if one wants to (via a configuration option for example).

> By now im strongly opposed to use type hints for parameter configuration, it's just terrifying complex to read

I'm leaning towards this as well.

I however find using *default values* for that (like `ward` does) to be quite readable:

```python
@pytest.fixture
def random_sample() -> List[int]:
    ...
    
def test_foo(x: List[int] = random_sample):
    ...
```

Or:

```python
def random_sample() -> List[int]:
    ...
    
def test_foo(x: List[int] = Fixture[random_sample]):
    ...
```

The second form has the advantage of not requiring to decorate the fixture function at all.

Also it seems you can make it work with `mypy`:

```python
from typing import List, TypeVar, Callable

T = TypeVar("T")

def Fixture(func: Callable[..., T]) -> T:
    ...

def random_sample() -> List[int]:
    return [1, 2]

def test_foo(x: List[int] = Fixture(random_sample)):
    x.append(10)
```

> Until we enable to actually do it, i don't see value in hypothesis on how it will look, as implementation and hitting issues will reshape the system.

Yeah I agree. First order of business would be to change the internals to allow this type of system to be provided by a plugin.

**Comment by shoyer** [comment]
> ```python
> def random_sample() -> List[int]:
>     ...
>     
> def test_foo(x: List[int] = Fixture[random_sample]):
>     ...
> ```

It's a little weird to see type annotation syntax with `[]` used for default values. Did you mean to write something like `test_foo(x: List[int] = Fixture(random_sample))` like what you wrote below?

**Comment by nicoddemus** [comment]
I did, thanks (corrected original post).

**Comment by graingert** [comment]
it doesn't seem that bad using `typing.Annotated`

```python
def random_sample() -> List[int]:
    ...
    
def test_foo(x: pytest.Fixture[List[int], random_sample]):
    ...
```

**Comment by graingert** [comment]
and class based fixtures work nicely too

```python
@attr.frozen
class RandomSample:
    v: List[int]

    @pytest.fixture
    @classmethod
    def fixture(cls):
        return cls([1, 2, 3])


def test_foo(y: RandomSample):
    x = y.v
```

**Comment by shoyer** [comment]
I would still lean towards a decorator for tests rather than type annotations or default values. Type annotations and default values would be more compact, but the decorator removes all the magic.

In principle, the core test collection part of pytest wouldn't need to know about fixtures at all anymore -- the `use_fixture` decorator could just be a normal decorator that returns a new test function. For example, if we use a decorator we wouldn't need to pre-define a "scope" for fixtures, and could just rely on Python's normal variable model, e.g.,
```python
def smtp_connection():
    import smtplib
    return smtplib.SMTP("smtp.gmail.com", 587, timeout=5)

@pytest.use_fixture(smtp_connection)  # local-scope
def test_ehlo(connection):
    ...

cached_smtp_connection = functools.lru_cache(smtp_connection)

@pytest.use_fixture(cached_smtp_connection)  # module-scope
def test_ehlo(connection):
    ...
```

**Comment by graingert** [comment]
> core test collection part of pytest wouldn't need to know about fixtures at all anymore

This seems like something that could be developed on pypi for use with even old versions of pytest

**Comment by ms-lolo** [comment]
> In principle, the core test collection part of pytest wouldn't need to know about fixtures at all anymore

this is absolutely what we should be striving for. eliminating any kind of magic and asking test writers to type a few extra characters will eliminate tight coupling between the pytest components and give us a much more flexible and maintainable architecture.

**Comment by graingert** [comment]
> asking test writers to type a few extra characters

test writers could also use pytestmark and a collect_item hook to add this decorator, similar to pytest-asyncio

**Comment by graingert** [comment]
> Could someone help me understand a little more about what current limitations live in pytest that prevent me from wrapping a normal test function in a decorator? I'd love to help poke at things a bit more and learn more about the pytest internals in the process.

@altendky maintains some workarounds for Twisted decorators, and there's workarounds for `@mock.patch`

**Comment by nicoddemus** [comment]
@shoyer

> Type annotations and default values would be more compact, but the decorator removes all the magic.

But that doesn't require *any* changes to pytest, that's something you can implement on your own today. 

Consider that `mock.patch` and `hypothesis.given` (just to cite a few examples) already work with pytest today without changing pytest at all.

@ms-lolo

> Could someone help me understand a little more about what current limitations live in pytest that prevent me from wrapping a normal test function in a decorator? I'd love to help poke at things a bit more and learn more about the pytest internals in the process.

There's none AFAIK. Here's a quick example of injecting custom fixtures using a normal Python decorator:

```python
def myfixtures(**input_fixtures):

    def wrapped(fn):

        def test_func():
            # setup fixtures
            fixture_gens = {name: fixture_func() for name, fixture_func in input_fixtures.items()}
            kwargs = {name: next(fixture_gen) for name, fixture_gen in fixture_gens.items()}
            try:
                fn(**kwargs)
            finally:
                for name, fixture_gen in fixture_gens.items():
                    try:
                        next(fixture_gen)
                    except StopIteration:
                        pass
                    else:
                        raise RuntimeError(f"{name} can only yield once")

        test_func.__name__ = fn.__name__
        return test_func

    return wrapped


def foo():
    print()
    print("foo setup")
    yield 42
    print("foo teardown")


@myfixtures(x=foo)
def test_foo(x):
    print("test_foo started")
    assert x == 42
    print("test_foo done")
```

Output:

```
 λ pytest .tmp\test_custom_fixtures.py -s --no-header
======================== test session starts ========================
collected 1 item

.tmp\test_custom_fixtures.py 
foo setup
test_foo started
test_foo done
foo teardown
.

========================= 1 passed in 0.02s =========================
```

**Comment by ms-lolo** [comment]
> There's none AFAIK. Here's a quick example of injecting custom fixtures using a normal Python decorator:

Well that looks promising! Then maybe I was just struggling with `parametrize` that has a similar string value being interpreted for special meaning 🤔 but your example is definitely helpful thank you!

**Comment by shoyer** [comment]
> > Type annotations and default values would be more compact, but the decorator removes all the magic.
> 
> But that doesn't require _any_ changes to pytest, that's something you can implement on your own today.

OK, awesome -- maybe we should just encourage that pattern!

There's at least a small amount of associated boilerplate (e.g., your `myfixtures` function), but that could certainly be prototyped in a separate library. Given how popular this idea seems to be, it might make sense to eventually move that into pytest as an alternative to pytest's standard fixtures?

**Comment by shoyer** [comment]
Here's my attempt to get this working with context managers:
```python
from contextlib import ExitStack, contextmanager

def use_fixtures(*args, **kwargs):
    def decorator(func):
        def wrapper():
            with ExitStack() as stack:
              args2 = [stack.enter_context(a()) for a in args]
              kwargs2 = {k: stack.enter_context(v()) for k, v in kwargs.items()}
              out = func(*args2, **kwargs2)
              return out
        wrapper.__name__ = func.__name__
        return wrapper
    return decorator


@contextmanager
def foo():
    print(f"foo setup")
    yield 42
    print("foo teardown")


@use_fixtures(x=foo)
def test_foo(x):
    print("test_foo started")
    assert x == 42
    print("test_foo done")


@contextmanager
@use_fixtures(x=foo)
def nested(x):
    print(f"nested setup")
    yield x
    print("nested teardown")

@use_fixtures(x=nested)
def test_nested(x):
    print("test_foo started")
    assert x == 42
    print("test_foo done")
```

`test_foo()` gives the right result:
```
>>> test_foo()
foo setup
test_foo started
test_foo done
foo teardown
```

But the nested context manager doesn't (`foo teardown` should happen at the end):
```
>>> test_nested()
foo setup
foo teardown
nested setup
test_foo started
test_foo done
nested teardown
```

I'm probably missing something obvious here...

**Comment by nicoddemus** [comment]
> Given how popular this idea seems to be, it might make sense to eventually move that into pytest as an alternative to pytest's standard fixtures?

Perhaps, but here we are only getting our feet wet: fixtures support scopes, parametrization, autousing, etc. For that to be included into the core, there's a long road ahead both in implementation and designing how those new pieces fit together with the rest of pytest's feature. 

Having said that, this is an excellent candidate for a plugin: you are free to experiment different designs at your leisure.

**Comment by techdragon** [comment]
> Perhaps, but here we are only getting our feet wet: fixtures support scopes, parametrization, autousing, etc. 

- Scopes feel like they are something that could be left to each individual fixture/callable... I've done things analogous to scopes for quite a few different use cases over the years, I'm pretty sure a cached frame (or other runtime introspection options) lookup would give more than enough context to make any of these scopes work... I have even gone one layer "deeper" on one particular occasion and generated a unique callable state/scope based on each line number and character position when I wrote this https://github.com/techdragon/python-flip-flop-operator/blob/master/src/flip_flop_operator/flip_flop.py 
  - _NOTE: Anyone reading that link, please don't judge the obvious madness of a flip flop operator without the context that prompted its creation. For that please watch this amusing conference lightning talk https://youtu.be/rNkbmu4e3MA?t=2667  My linked code there was the "winner"_ :sweat_smile:
- Re-reading the parameterisation docs, isn't this effectively just a mechanism to unpack otherwise normal fixtures? Thus we get a special "parameterized" equivalent of `use_fixture` that takes a callable like our normal `use_fixture` will, and that callable will return/yield one-or-more callables , which are then passed into each iteration of the test... Much like the current test collection machinery does, just more explicitly.
- personally I see "autouse" fixtures and "explicit" fixtures, a.k.a. "fixtures without name matching magic", to be philosophically opposed concepts. The idea of trying to make fixtures more explicit requires abandoning the notion of autouse fixtures as a feature. The docs themselves warn against potential adverse behaviour of this feature used without care, "please use that with care: if you define an autouse fixture in a plugin, it will be invoked for all tests in all projects where the plugin is installed. " 
  - I would rather explicitly use the fixture decorator on every method.
  - Or if I _really_ wanted that kind of behaviour, I'd consider writing a MetaClass like this https://stackoverflow.com/a/11350487/1950432 
  - Or if its something large like starting/stopping a database or service necessary for a subset of the test suite ... I'd probably gather them into a module (or maybe mark them) and put that in its own `tox` environment... Let `tox` take care of ensuring I have an environment suitable for those tests not my test runner.

**Comment by darrenburns** [comment]
@nicoddemus One thing to note about the default arg approach is that it breaks things like hypothesis, which doesn't permit default arguments. Not sure if that's an issue, but thought I'd raise it. That's one of the reasons why Ward supports both the "default argument" approach and the decorator approach.

e.g. you can do
```python
@fixture
def my_fixture():
    yield 1

@test("check that it is 1")
def _(f=my_fixture):
    assert f == 1
```

OR, using the `@using` decorator as has also been suggested here (and seems to be the preferred approach):

```python
@fixture
def my_fixture():
    yield 1

@test("check that it is 1")
@using(f=my_fixture)
def _(f):
    assert f == 1
```

**Comment by nicoddemus** [comment]
Ahh right, thanks @darrenburns!

**Comment by bkhl** [comment]
> OR, using the `@using` decorator as has also been suggested here (and seems to be the preferred approach):

That looks helpful, but I don't see where `@using` comes from, and it turns out to be quite hard to search for. Can anyone point to where to import that from?

**Comment by burnpanck** [comment]
What is the current status of this?

**Comment by nicoddemus** [comment]
The current status is basically: not something that will be implemented in the core directly, but which someone can experiment using a plugin. Not sure if any progress has been done on the latter.

Btw is this useful to leave open?

**Comment by burnpanck** [comment]
I think it is. There are quite few people out there who would love such a feature (including prominent people such as @shoyer). This issue can serve as a meeting point to attempt pushing each other to move this a bit further ;-).

Magic name matching is a constant hurdle for new contributors not familiar with any given code-base: If a fixture is missing, there is absolutely no indication in the code where it is supposed to be loaded from - is it a local `conftest.py`, a missing pytest plugin or on some other obscure path. IMHO it just feels against the principles of python (e.g. 12. "In the face of ambiguity, refuse the temptation to guess." or 19. "Namespaces are one honking great idea -- let's do more of those!"). 

I guess the problem is that one usually gets bitten by this while working with foreign packages rather than one's own.

**Comment by nicoddemus** [comment]
Thanks @burnpanck. 

Even if closed, the issue is still searchable, and people can further comment. My experience is that by leaving it open we are signalling that this is something we plan to integrate into the core eventually, and people then ask the status on it because (understandably) don't want to go through dozens of posts to understand the status.

But I don't think it is a big deal to leave it open too.

**Comment by RonnyPfannschmidt** [comment]
Personally, ever since typing and type annotation is part of the language I was hoping to experiment with this,

Unfortunately it's not clear if/when we can even create a base for this

**Comment by adriangb** [comment]
Hi folks, I've been working on a dependency injection system ([di](https://github.com/adriangb/di)) that has an API inspired by FastAPI and scoping/caching rules (somewhat) inspired by Pytest. @graingert thought it might be applicable to this issue, so I wrote up a quick prototype:


```python
from typing import AsyncGenerator, List
import pytest

from di.pytest import Depends, inject


class Log(List[str]):
    pass


def dep(log: Log = Depends(Log)) -> int:
    log.append("dep")
    return 1


async def generator_dep(inner_val: int = Depends(dep), log: Log = Depends(Log)) -> AsyncGenerator[int, None]:
    log.append("generator_dep setup")
    yield inner_val + 1
    log.append("generator_dep teardown")


@pytest.mark.anyio
@inject
async def test_inject(v: int = Depends(generator_dep), log: Log = Depends(Log)) -> None:
    assert v == 2
    assert log == ["dep", "generator_dep setup"]  # teardown hasn't run yet
```

My WIP branch for this is here: https://github.com/adriangb/di/pull/43

Before I do any more work on it, I wanted to post here to see if there'd be any interest / conceptual input into what an integration like this should look like. I have some thoughts on rough edges but I'll hold off on them for now.

**Comment by grahamgower** [comment]
I've been trying to annotate all my tests that use a fixture with `pytest.mark.usefixtures` (I don't write my own fixtures, because I dislike the pattern so much). Would it be possible to have a configuration option to disable magic name matching without an explicit `pytest.mark.usefixtures`?

**Comment by RonnyPfannschmidt** [comment]
@grahamgower that particular one is out of scope for this issue and currently extremely unlikely as currently pytest has only name matching


@adriangb that approach looks interesting, but unless there is a way to use annotations instead of magic default values, this doest sit too well with me 

with the current tech debt in pytest, im also under the impression, that unless we slowly externalize our own system first, we are unlikely to have success with integrating a new system in a way that wont break immense amounts of test-suites

in particular parameterization, dependencies and setup reduction reordering are going to be very challenging topics

i absolutely want pytest to see a better DI system, however we have to evolve the codebase into using one

if there was a api we could in parts start to use for better management of setupstate, value storage, interacting with fixture definition,
i beleive that would be a much better starting point than the outer facade of a di system

as pytest has a already existing internal DI system thats by no means easy to replace/reshape

**Comment by adriangb** [comment]
> @adriangb that approach looks interesting, but unless there is a way to use annotations instead of magic default values, this doest sit too well with me

could you clarify what you mean? We also support `v: Annotated[int, Depends(generator_dep)]`, and when you're injecting a class you don't even need that (`log: Log`). This is covered [here](https://www.adriangb.com/di/0.33.0/wiring).

My idea was to interface with the existing DI system. You'd be able to opt in to the above proposed DI system by (installing the dependency) and then adding the `@inject` decorator (or something like that). The test function itself would still be able to accept pytest fixtures, but sub-dependencies would not (so limited interoperability). And yeah you'd probably have to rewrite a lot of your code to really take advantage of it, but that would still be the case if pytest came out with a feature that allowed you to match on fixture objects/values instead of names.

**Comment by RonnyPfannschmidt** [comment]
@adriangb i like that annotation approach, for pytest i would probalby have a helper to make those annotations spell nicely 

its pretty strange that mypy/pyrright "interpret" strange default values, but don't handle annotations nicely

i suspect there will be some extra need for plugins (just like attrs, pydantic & co)

**Comment by adriangb** [comment]
> i would probalby have a helper to make those annotations spell nicely

Do you mean like an `Annotated` alias?

> its pretty strange that mypy/pyrright "interpret" strange default values, but don't handle annotations nicely

to be fair to those tools, [it's a hack](https://github.com/adriangb/di/blob/36c757374119b7ac6d30adf27e4f0b0bbee0ee01/di/params.py#L79) (same thing FastAPI does more or less)

> i suspect there will be some extra need for plugins (just like attrs, pydantic & co)

a plugin for MyPy or for pytest?

**Comment by graingert** [comment]
@RonnyPfannschmidt 

> but don't handle annotations nicely

what do you mean by that?

**Comment by RonnyPfannschmidt** [comment]
@graingert thats related to my misunderstanding about having the `Dependent` global function just ignore the return value for type checking

as it turns out, the "support" is just a tactical type ignore, no actual support is there

**Comment by rtaycher** [comment]
> 
> There's none AFAIK. Here's a quick example of injecting custom fixtures using a normal Python decorator:
> 
> ```python
> def myfixtures(**input_fixtures):
> 
>     def wrapped(fn):
> 
>         def test_func():
>             # setup fixtures
>             fixture_gens = {name: fixture_func() for name, fixture_func in input_fixtures.items()}
>             kwargs = {name: next(fixture_gen) for name, fixture_gen in fixture_gens.items()}
>             try:
>                 fn(**kwargs)
>             finally:
>                 for name, fixture_gen in fixture_gens.items():
>                     try:
>                         next(fixture_gen)
>                     except StopIteration:
>                         pass
>                     else:
>                         raise RuntimeError(f"{name} can only yield once")
> 
>         test_func.__name__ = fn.__name__
>         return test_func
> 
>     return wrapped
> 
> 
> def foo():
>     print()
>     print("foo setup")
>     yield 42
>     print("foo teardown")
> 
> 
> @myfixtures(x=foo)
> def test_foo(x):
>     print("test_foo started")
>     assert x == 42
>     print("test_foo done")
> ```
> 
> Output:
> 
> ```
>  λ pytest .tmp\test_custom_fixtures.py -s --no-header
> ======================== test session starts ========================
> collected 1 item
> 
> .tmp\test_custom_fixtures.py 
> foo setup
> test_foo started
> test_foo done
> foo teardown
> .
> 
> ========================= 1 passed in 0.02s =========================
> ```
> 

Is there a way to do this in pytest > 4

I just get an error for

Fixture ""called directly. Fixtures are not meant to be called directly . .. and some links to the deprecation in the online docs

**Comment by rtaycher** [comment]
@nicoddemus 

your `myfixtures` decorator fixes the problem I have with pytest, making it more explicit. 
Sadly it seems to not work anymore due to the deprecation and removal of directly calling fixtures

https://docs.pytest.org/en/7.1.x/deprecations.html?highlight=calling#calling-fixtures-directly (
Removed in version 4.0)

Are there any ways to make it work again with branch 7.2.*?

**Comment by nicoddemus** [comment]
I just tried the example from https://github.com/pytest-dev/pytest/issues/3834#issuecomment-685690616 and it works for me.

That error only happens when someone uses the `@fixture` decorator, which the example does not.

One workaround is to extract the fixture code you have into a normal function, and call that from the fixture and the `myfixtures` code.

**Comment by rtaycher** [comment]
Yes I'm sorry if I confused you.
I used the fruit basket example from the pytest documentation which has a fixture decorator.

My understanding is that the fixture decorator makes it into a "proper" fixture and all the third party pytest plugins use such fixtures. I was hoping for a workaround for pytests name based matching which I worry will confuse people on my team. Maybe I misunderstood what you were trying to say in that comment.

Is there a way to do make a decorator similar to what you wrote possibly with getfixturevalue or something else that works on fixture decorated functions or am I just totally misunderstanding how the fixtures work and that isn't currently possible without a lot of effort?

**Comment by rtaycher** [comment]
This seems to work as a useful workaround
```python
def make_test_with_fixtures_from_defaults(fn):
    def default_func_names(func):
        # HT: https://stackoverflow.com/a/12627202/259130
        signature = inspect.signature(func)
        return [
            v.default.__name__
            for k, v in signature.parameters.items()
            if v.default is not inspect.Parameter.empty
        ]

    fixture_names = default_func_names(fn)

    z = textwrap.dedent(f'''\
    def test_func({', '.join(fixture_names)}):
        return fn({', '.join(fixture_names)})
    ''')
    print(z)
    scope = dict(fn=fn, fixture_names=fixture_names)
    exec(z, scope)
    test_func = scope['test_func']
    test_func.__name__ = fn.__name__
    return test_func
```
seems to work. Assuming you don't call the function by keyword argument. Does that every happen? Or are there any other corner cases or plugins this might have trouble with?

```python
@make_test_with_fixtures_from_defaults
def test_create_file(tmp=tmp_path):
    d = tmp / "sub"
    d.mkdir()
    p = d / "hello.txt"
    p.write_text('lol')
    assert p.read_text() == 'lol'
    assert len(list(tmp.iterdir())) == 1
```
works as expected and passes(so do other examples I tried)

I know a lot of people don't like exec but it works. I could probably do something to set the signature with https://smarie.github.io/python-makefun/ instead but I think makefun might also use exec. Maybe https://github.com/dfee/forge?

**Comment by nicoddemus** [comment]
Interesting approach. If you turn that into a package, I'm sure others would be interested in trying it out.

**Comment by rtaycher** [comment]
I think this version of the workaround should support both default argument and decorator provided fixtures like ward does
```python
def make_test_with_fixtures(**fixture_kwargs):
    def inner(fn):
        signature = inspect.signature(fn)
        keyword_names_to_fixtures = {
            k: fixture_kwargs.get(k, None) or v.default
            for k, v in signature.parameters.items()
        }
        assert all(v is not inspect.Parameter.empty
                   for v in keyword_names_to_fixtures.values()), (
            'every parameter should have a matching fixture function '
            'provided in either the decorator or default function')
        keyword_names_to_fixture_names = {k: f.__name__ for (k, f) in keyword_names_to_fixtures.items()}
        fixture_names = keyword_names_to_fixture_names.values()

        z = textwrap.dedent(f'''\
        def test_func({', '.join(fixture_names)}):
            return fn({', '.join(kname+'='+fname for (kname, fname) in keyword_names_to_fixture_names.items())})
        ''')
        scope = dict(fn=fn)
        exec(z, scope)
        test_func = scope['test_func']
        test_func.__name__ = fn.__name__
        return test_func
    return inner
```
I tested it out and both
```

@make_test_with_fixtures()
def test_bar1(f1=fix_w_yield1, f2=fix_w_yield2, tmp=tmp_path):
    print("test_bar")
    print(f'{tmp}')
    assert tmp.exists()

@make_test_with_fixtures(f1=fix_w_yield1, f2=fix_w_yield2, tmp=tmp_path)
def test_bar2(f1, f2, tmp):
    print("test_bar")
    print(f'{tmp}')
    assert tmp.exists()
```
test_bar1 and test_bar2 seem to work.

**I'll try to put it in a package when I have time. Maybe this weekend**

I want to see it it possible avoid exec first(possibly via the python-forge package).

**Comment by nicoddemus** [comment]
> I want to see it it possible avoid exec first(possibly via the python-forge package).

Note that is an implementation detail, which you can change at will later. 👍

**Comment by rtaycher** [comment]
It's pretty barebones but people might find it useful to reference fixtures by reference.

https://github.com/rtaycher/pytest-fixture-ref
pip install git+https://github.com/rtaycher/pytest-fixture-ref.git
people should feel free to make suggestions or PRs or suggest name changes
I'll try to publish it on pypi before the weekend

pip install git+https://github.com/rtaycher/pytest-fixture-ref.git
```python
from pytest_fixture_ref import make_test_with_fixtures

@make_test_with_fixtures()
def test_bar1(f1=fix_w_yield1, f2=fix_w_yield2, tmp=tmp_path):
    print("test_bar")
    print(f'{tmp}')
    assert tmp.exists()


@make_test_with_fixtures(f1=fix_w_yield1, f2=fix_w_yield2, tmp=tmp_path)
def test_bar2(f1, f2, tmp):
    print("test_bar")
    print(f'{tmp}')
    assert tmp.exists()
```

**Comment by rtaycher** [comment]
@nicoddemus is there a common way to import pytest plugins? 
`from _pytest.tmpdir import tmp_path` works for tmp_path but I couldn't find some of the others.
What about various pytest plugins?

**Comment by nicoddemus** [comment]
There is no standard for that, each plugin decides which API to expose; so in your case `from pytest_fixture_ref import make_test_with_fixtures` is fine. :+1: 

Btw, if you have Twitter, feel free to tweet about that project and I'll be glad to retweet (also we can ask the pytest-dev official twitter to retweet too).

**Comment by ms-lolo** [comment]
@rtaycher can I suggest splitting up the two versions of fixtures into their own decorators? I'm opposed to the version that inspects the functions' default values and I think it would be useful to have that decision be clearer when we write our tests. A lot of development on large code bases happens by copying nearby code and having them as separate decorators makes it clearer that, in this code base, we chose style A over style B.

**Comment by rtaycher** [comment]
> @rtaycher can I suggest splitting up the two versions of fixtures into their own decorators? I'm opposed to the version that inspects the functions' default values and I think it would be useful to have that decision be clearer when we write our tests. A lot of development on large code bases happens by copying nearby code and having them as separate decorators makes it clearer that, in this code base, we chose style A over style B.

Could you open an issue with some suggested names? I tend to bikeshed maybe @fixtures_from_default_value and @fixtures_from_ref? Both sound not quite right to me

**Comment by rtaycher** [comment]
@nicoddemus https://twitter.com/rtaycher1987/status/1593199919892156418
@ms-lolo I split them into using_fixtures_from_defaults/using_fixtures_from_kwargs since I realized this would work not just for tests but for fixtures that take other fixtures

sorry it took so long. Figuring out the CI/publish setup for the boilerplate I used took a while(hit a number of odd library issues, plus the autochangelog is really annoying)

**Comment by nicoddemus** [comment]
Nice, thanks @rtaycher!

I also recommend that suggestions/requests to be made in https://github.com/rtaycher/pytest-fixture-ref/issues instead of in this thread, as that is the appropriate place.  :+1:

**Comment by YakovL** [comment]
A couple of notes:

1. This also hurts DX if one uses Pylance as it also complains like Pylint
2. The workaround

   ```
   @pytest.fixture(name='session_manager')
   def get_session_manager():
       return GameSessionsManager()

   def test_create_session(session_manager: GameSessionsManager):
       ...
   ```
   is fragile: since the `GameSessionsManager` type is casted to the argument, once implementation of `get_session_manager` is changed (say, returns `GameSessionsManager | None`), things can get broken.

**Comment by ZeeD** [comment]
@adriangb your proposal sounded very promising 
Did the work continued ? Is there some branch of pytest that could be used?

**Comment by adriangb** [comment]
I got bogged down in internal implementations of pytest and never got to finishing it. It mostly worked, and worked well, but there was bugs I just couldn't fix. I don't think I'll have time to do it myself.

**Comment by RonnyPfannschmidt** [comment]
can you provide a open draft pr for reeference - it may enable us to put focus on the bugs to enable the rest

**Comment by adriangb** [comment]
I think this was my WIP branch: https://github.com/adriangb/di/tree/pytest. But given that it was last touched 3 years ago I don't know the state of it. To be clear I'm not saying there were bugs in pytest, more so that I just didn't understand how things were supposed to work internally, especially when it came to the async plugins.

**Comment by tristan** [comment]
I have thrown together a quick plugin that allows fixtures to be resolved using `Annotated[...]` rather than their full name.

e.g.
```
from typing import Annotated
from pytest_annotated import Fixture

class SomeThing:
    pass

@pytest.fixture
async def somefixture(someotherfixture) -> SomeThing:
    return SomeThing()

async def test_thing(st: Annotated[SomeThing, Fixture(somefixture)]):
    assert isinstance(st, SomeThing)
```

you can find it here: https://github.com/tristan/pytest-annotated and here: https://pypi.org/project/pytest-annotated/

Unfortunately, I feel it requires the use of too many `_private` parts of pytest, so I'm a bit scared of the maintenance burden it will require to keep it up to date. Would be super happy to have someone more familiar with the internals of pytest to make suggestions for making it less brittle .

**Comment by RonnyPfannschmidt** [comment]
that one is a funky little hack, we ought too figure if we can enable lookup by name, type and annotated type in future

**Comment by millerdev** [comment]
https://pypi.org/project/pytest-unmagic/ is a pytest plugin that implements fixtures with with normal import semantics.

Fixtures are applied to tests or other fixtures with a `@use` decorator. Alternately, a fixture can be called without arguments to get its value inside a test or other fixture function.

**Comment by RonnyPfannschmidt** [comment]
@millerdev at first glance pytest-unmagic is a complete deception - using magic globals unrelated to the actual fixtures instead of actual values passed - i strongly discourage anyone from using it - it does the opposite of its name

**Comment by millerdev** [comment]
@RonnyPfannschmidt Fixtures are global objects. Pytest registers them in an opaque hierarchical namespace, and effectively does a global lookup each time a fixture is requested. The magic that this plugin eliminates is function-scoped local names magically invoking global fixtures. Granted, the active request reference is not the nicest feature. Hopefully we can work together to improve it?

The active request reference allows `unmagic.fixture` to work with pytest features such as `--setup-show` and `--fixtures`. While implemented as a module-level singleton by default, it does provide a mechanism for thread-local or other custom scoping. In most cases this will be unnecessary.

I have a development branch where the global active request reference has been removed, but it has trade-offs. The implementation is more complicated, and it is impossible to use fixtures in the context of `unittest.TestCase`. I didn't release it (yet) because the test suite that pytest-unmagic was originally intended for uses unittest-style tests extensively, and allowing fixtures to be used there is a big win.

For now—in the very spirit of pytest fixtures—practicality beats purity.

**Comment by nicoddemus** [comment]
My 2 cents on the plugin -- I agree with @RonnyPfannschmidt overall about the design, however I'm also happy that @millerdev did an experiment to try things out (it is his project after all!).

But regardless I wanted to share my opinions on some points (I'm not speaking for all maintainers or for the pytest project at all, but only how I see this topic):

* The existence of an external plugin that does "fixtures without name matching" does not automatically imply endorsement from the pytest project, much less that it will be incorporated at some point.

* I think pytest should never stray from the current fixture mechanism, or even support both, even if there are people clamoring for that to be the case -- besides being an controversial topic[^1], there's backward compatibility, and having to teach/learn multiple ways of doing the same thing.

* I **do think**, however, that we should improve pytest so *plugins can change how fixtures are handled*, allowing plugins to implement whatever mechanism one prefers over the current name matching in core pytest.


[^1]: I think the current fixture mechanism in pytest is absolutely fine, IMO lots of frameworks implement things that seem like "magic" at first glance, for example dataclasses "magically" injecting methods, SQL/Django columns  "magically" talking to a DB based on how attributes are declared... if you squint hard enough many framework features can be shrugged off (or even bad mouthed) as magic.

**Comment by burnpanck** [comment]
@nicoddemus: Clarke's third law states ["any sufficiently advanced technology is indistinguishable from magic"](https://en.wikipedia.org/wiki/Clarke%27s_three_laws). That is not at all a bad thing. It just means that as soon as a mechanism is not obvious how it works, it appears as if it were magic. Once you understand how things work, its not magic anymore, but just a great power to wield. And with great power, comes great responsibility.

IMHO, what is wrong with the current fixture mechanism is the _single global namespace_ shared among all participants in a given test session. Thus, this particular great power is hard to guide to its intended target. Rather than a surgical laser beam, these fixtures behave like a virus; once you unleash them, they go everywhere, hitting both intended and unintended targets. It appears magic to the many users, because once that virus arrives at your site (intended or not), you don't see where it comes from. This applies both to humans and to tools (type checkers). More importantly, the design makes it _impossible_ to reason about the area of effect within any subscope, because you always need to know the full testing session before you can start figuring out where fixtures may come from or go to.

This is very different from the magic methods in dataclasses or the various forms of automatic schema validation and column creation in ORMs. Its true that it can appear equally magic to someone who doesn't understand the mechanism, but once you do understand, a) the rules are simple (as in the current pytest fixtures) and b) the amount of inputs you need to consider to apply them are reasonably bounded (unlike in the current pytest fixtures).

**Comment by nicoddemus** [comment]
Thanks @burnpanck for the detailed reasoning, you make good points.

Regardless, all my previous points still stand: backward compatibility, more than one way to do the same thing, etc, and those do not change regardless of the discussion about what constitutes a magical behavior or not.

But again, this is all from my POV, I don't know if other maintainers feel otherwise and think it is OK to add an additional way to use fixtures to core pytest.

**Comment by millerdev** [comment]
> **I do think**, however, that we should improve pytest so _plugins can change how fixtures are handled_, allowing plugins to implement whatever mechanism one prefers over the current name matching in core pytest.

Yay! :tada: This would be wonderful.

> what is wrong with the current fixture mechanism is the _single global namespace_ shared among all participants

This is precisely why I think pytest-unmagic is an improvement on pytest magic fixture parameters. "Unmagic" fixtures are always explicitly referenced, and do not occupy a single shared global namespace. The question "where is the code for that?" can be answered without any knowledge of pytest fixture name resolution.

The part that @RonnyPfannschmidt reacted so strongly to (I think) is the active request reference, which is an implementation detail that allows the plugin to bind fixture setup and teardown to the appropriate scope and to retrieve the value from pytest's internal fixture cache. It is easy to reason about the active request within the context of any given test or fixture, so that single global reference does not seem as problematic as a shared global namespace for all fixtures.

Maybe "unmagic" is a bit of a misnomer. The plugin moves the "magic" part from fixture name resolution, which everyone using fixtures is confronted with, to an implementation detail where most people don't need to know much if anything about it. You might need to know if you're running tests in parallel within a single Python interpreter, but that is not a problem I tried to solve in v1.0.

**Comment by septatrix** [comment]
I think the annotation way which was already raised there is a very good idea. It is not that magical (pretty much just using `typing.get_type_hints(metafunc.definition.function)` + a fixture registry) and has already reached significant adoption by projects like SQLAlchemy and FastAPI. The implementation by @tristan https://github.com/tristan/pytest-annotated already looks quite good. It would be great to either open the private APIs which he is currently using or consider incorporating it into core pytest.

Regarding two ways to archive the same thing I also do not really have concerns. There are already other aspects where two approaches are used in pytest. The one that directly comes to my mind is `import-mode` which has gained `importlib` and is now recommended.

**Comment by adriangb** [comment]
+1 for the approach `pytest-annotated` seems to be taking!

**Comment by RonnyPfannschmidt** [comment]
pytest-annotated looks like a good prototype 

However we are not in a position to open the apis it's using 

But I'm more than happy with working towards integrating types with fixture lookup
